zipline/assets/assets.py

```diff
@@ -526,33 +538,38 @@ def map_identifier_index_to_sids(self, index, as_of_date):
     def _compute_asset_lifetimes(self):
         """
         Compute and cache a recarry of asset lifetimes.
-
-        FUTURE OPTIMIZATION: We're looping over a big array, which means this
-        probably should be in C/Cython.
         """
-        with self.conn as transaction:
-            results = transaction.execute(
-                'SELECT sid, start_date, end_date from equities'
-            ).fetchall()
-
-            lifetimes = np.recarray(
-                shape=(len(results),),
-                dtype=[('sid', 'i8'), ('start', 'i8'), ('end', 'i8')],
-            )
-
-            # TODO: This is **WAY** slower than it could be because we have to
-            # check for None everywhere.  If we represented "no start date" as
-            # 0, and "no end date" as MAX_INT in our metadata, this would be
-            # significantly faster.
-            NO_START = 0
-            NO_END = np.iinfo(int).max
-            for idx, (sid, start, end) in enumerate(results):
-                lifetimes[idx] = (
-                    sid,
-                    start if start is not None else NO_START,
-                    end if end is not None else NO_END,
-                )
-        return lifetimes
+        equities_cols = self.equities.c
+        buf = np.array(
+            tuple(map(
+                float,
+                sa.select((
+                    equities_cols.sid,
+                    equities_cols.start_date,
+                    equities_cols.end_date,
+                )).execute(),
+            )),
+            dype='<f8',  # use doubles so we get NaNs
```

Comment:
Yes.  Or better yet, have those column be non-nullable and use defaults on insertion.